lib/tree.rb

```diff
@@ -16,42 +16,136 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    # RECURSIVE WAY 
+      @root = add_helper(@root, key, value)
+
+    # LOOP WAY
+    # if @root.nil? 
+    #   @root = TreeNode.new(key, value)
+    # else 
+    #   current = @root 
+    #   while true 
+    #     if key < current.key 
+    #       if !current.left.nil? 
+    #         current = current.left
+    #       else 
+    #         current.left = TreeNode.new(key, value)
+    #       end 
+    #     else 
+    #       if !current.right.nil? 
+    #         current = current.right
+    #       else 
+    #         current.right = TreeNode.new(key, value)
+    #       end 
+    #     end 
+    #   end 
+    # end 
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # helper method for add() recursive way 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil? 
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else 
+      current_node.right = add_helper(current_node.right, key, value)
+    end 
+
+    return current_node
+  end 
+
+
+  # Time Complexity: o(log n)
+  # Space Complexity: 0(n)
   def find(key)
```

Comment:
The space complexity is O(log n) if the tree is balanced and O(n) if it's not.
Otherwise 👍